src/main/python/ayab/engine/engine.py

```diff
@@ -46,28 +48,31 @@ class Engine(SignalSender, QDockWidget):
     """
     port_opener = Signal()
 
-    def __init__(self, parent):
+    pattern:Pattern
+    status:StatusTab
+    
+    def __init__(self, parent:GuiMain):
         # set up UI
         super().__init__(parent.signal_receiver)
         self.ui = Ui_Dock()
         self.ui.setupUi(self)
-        self.config = OptionsTab(parent)
+        self.config:OptionsTab = OptionsTab(parent)
         self.config.portname = self.__read_portname()
         self.config.refresh()
         self.status = StatusTab()
         self.setup_ui()
         parent.ui.dock_container_layout.addWidget(self)
 
-        self.pattern = None
+        self.pattern:Pattern = None #type:ignore
```

Comment:
We _could_ type this as `Optional[Pattern]` and not have to have the ignore here, but that makes things everywhere else more annoying. Though maybe we should do that anyways because then we'd capture more edge cases.